cmd/release-controller/prune_graph.go

```diff
@@ -0,0 +1,93 @@
+package main
+
+import (
+	"bytes"
+	releasecontroller "github.com/openshift/release-controller/pkg/release-controller"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	kv1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/klog"
+)
+
+func (c *Controller) PruneGraph(secretClient kv1core.SecretInterface, ns, name string, printOption string, confirm bool) error {
+	stopCh := wait.NeverStop
+
+	releasecontroller.LoadUpgradeGraph(c.graph, secretClient, ns, name, stopCh)
+
+	imageStreams, err := c.releaseLister.ImageStreams(ns).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	var stableTagList []string
+
+	for _, imageStream := range imageStreams {
+		r, ok, err := releasecontroller.ReleaseDefinition(imageStream, c.parsedReleaseConfigCache, c.eventRecorder, *c.releaseLister)
+		if err != nil || !ok {
+			continue
+		}
+		if r.Config.As == releasecontroller.ReleaseConfigModeStable {
+			for _, tag := range imageStream.Spec.Tags {
+				stableTagList = append(stableTagList, tag.Name)
+			}
+		}
+	}
+
+	// To tags that are not present in the stable tags
+	toMissingList := make([]string, 0, len(c.graph.To))
+	for tag := range c.graph.To {
+		if !releasecontroller.StringSliceContains(stableTagList, tag) {
+			toMissingList = append(toMissingList, tag)
+		}
+	}
+
+	// From tags that are not present in the stable tags
+	fromMissingList := make([]string, 0, len(c.graph.From))
+	for tag := range c.graph.From {
+		if !releasecontroller.StringSliceContains(stableTagList, tag) {
+			fromMissingList = append(fromMissingList, tag)
+		}
+	}
+
+	// To tags that are not in the From tags
+	pruneTagList := make([]string, 0, len(c.graph.To))
+	for _, tag := range toMissingList {
+		if !releasecontroller.StringSliceContains(fromMissingList, tag) {
+			pruneTagList = append(pruneTagList, tag)
+		}
+	}
+
+	// From tags that are not in the To tags
+	for _, tag := range fromMissingList {
+		if !releasecontroller.StringSliceContains(toMissingList, tag) {
+			pruneTagList = append(pruneTagList, tag)
+		}
+	}
+
+	// Tags that are in both the From and To lists
+	for _, tag := range toMissingList {
+		if releasecontroller.StringSliceContains(fromMissingList, tag) {
+			pruneTagList = append(pruneTagList, tag)
+		}
+	}
```

Comment:
Is this just being done to combine the 2 lists without duplicates? The 1st and 3rd loops could be combined into:

```
pruneTagList = append(pruneTagList, toMissingList...)
```

Using sets could simplify this entire section into 1 line (plus an import):

```
pruneTagList := sets.NewString(append(toMissingList, fromMissingList...)...).UnorderedList()
```

Reply:
Done!